### PR TITLE
feat: refactor hgvsTools and support `c.` translation

### PR DESCRIPTION
**Pull request by larrybabb**
Address #365 

Begin refactoring the Translator logic to aggregate the functions and logic related to `hgvs` and provide support for `c.` from and to translation.  This should be step in the direction of re-thinking the organization of how the Translators are architected.

### CODE REVIEW

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 1]
```diff
@@ -1,26 +1,143 @@
...
 
 class HgvsTools():
+    """
+    A convenience class that exposes only the tools needed by vrs-python for working with HGVS (Human Genome Variation Society) notation.
 
-    """ A utility class that provides access to hgvs components that have a UTA database dependency"""
-    def __init__(self, uta_conn):
+    Attributes:
+        parser (hgvs.parser.Parser): The HGVS parser object.
+        uta_conn: The UTA (Unified Transcriptome and Annotation) connection object.
```
Comment (jsstevenson) [inline src/ga4gh/vrs/utils/hgvs_tools.py:18]: nitpick: it's "Universal Transcript Archive"

File: src/ga4gh/vrs/extras/translator.py [hunk 2]
```diff
@@ -64,39 +56,49 @@ def __init__(
...
+            "refseq:NG_": "g",
+            "refseq:NC_00": "g",
+            "refseq:NW_": "g",
+            "refseq:NR_": "n",
+            "refseq:NP_": "p",
+            "refseq:XM_": "n",
+            "refseq:XR_": "n",
+            "refseq:XP_": "p",
+            "GRCh": "g",
+        }
```
Comment (jsstevenson) [inline src/ga4gh/vrs/extras/translator.py:97]: double checking: this mapping isn't available somewhere else? I did a quick search through HGVS and didn't see it, and I wouldn't expect it to be in bioutils, but just making sure

File: src/ga4gh/vrs/dataproxy.py [hunk 3]
```diff
@@ -88,6 +89,56 @@ def translate_sequence_identifier(self, identifier, namespace=None):
...
+        actual_ref = self.get_sequence(sequence_id, start_pos, end_pos)
+        is_valid = actual_ref == ref
+        err_msg = ""
+        if not is_valid:
+            err_msg = (
+                f"Expected reference sequence {ref} on {sequence_id} at positions "
+                f"({start_pos}, {end_pos}) but found {actual_ref}"
+            )
+            _logger.warning(err_msg)
+        return is_valid, err_msg
```
Comment (jsstevenson) [inline src/ga4gh/vrs/dataproxy.py:141]: I think the return object could be simplified here. If `is_valid` is always false when `err_msg` is anything other than an empty string, and always true if `err_msg` is an empty string, does it add any additional information? Could we rephrase this to simply check for invalidity, returning `None` if there are no invalidations, and a string description if it is invalid? Alternatively -- maybe raise an exception if invalid and do nothing if valid?

**Comment by jsstevenson** [comment]
It would be nice, albeit out of scope here, if the generic `dataproxy` interface that's redefined here from seqrepo was located in a more central package so that we don't do any damage by making changes in VRS-Python without mirroring them back to SeqRepo.

File: src/ga4gh/vrs/extras/translator.py (hunk 2 above)
Comment (larrybabb) [reply src/ga4gh/vrs/extras/translator.py:97]: I know. it seems like someone stuck this here to make sure we only worked with "refseq" accessions. I do think that hgvs should retain this mapping structure since it is really an hgvs thing.

File: src/ga4gh/vrs/extras/translator.py [hunk 4]
```diff
@@ -64,39 +56,49 @@ def __init__(
...
+        str or None: The sequence type associated with the accession string, or None if no matching prefix is found.
+        """
 
-    def _get_hgvs_refget_ac(self, sv: hgvs.sequencevariant.SequenceVariant):
-        """Get refget accession for hgvs sequence variant"""
-        # prefix accession with namespace
-        sequence = coerce_namespace(sv.ac)
-        return self._get_refget_accession(sequence)
+        prefix_dict = {
+            "refseq:NM_": "n",
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:86]: I think this should be "c"
```suggestion
            "refseq:NM_": "c",
```

Comment (larrybabb) [reply src/ga4gh/vrs/extras/translator.py]: on it. I ended up moving some of this code to the seqrepo dataproxy classes.

File: src/ga4gh/vrs/extras/translator.py (hunk 2 above)
Comment (larrybabb) [reply src/ga4gh/vrs/extras/translator.py]: this was moved to the dataproxy classes for now. further refactoring will need to happen in another round to allow us to better organize/segregate hgvs code from vrs-python code.

File: src/ga4gh/vrs/utils/hgvs_tools.py (hunk 1 above)
Comment (larrybabb) [reply src/ga4gh/vrs/utils/hgvs_tools.py]: fixed

File: src/ga4gh/vrs/dataproxy.py (hunk 3 above)
Comment (larrybabb) [reply src/ga4gh/vrs/dataproxy.py:171]: i like your change request here. I am going to make a separate issue for it so as to not open another thread on this PR that is already bigger than I had anticipated.

**Comment by korikuzma** [comment]
@larrybabb can you pull in changes from `main` branch and resolve conflicts first?

File: notebooks/variation_normalizer_rest_dp.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
 import requests
 
+""" TODO convert this to a jupyter notebook """
+
```
Comment (korikuzma) [inline notebooks/variation_normalizer_rest_dp.py:4]: We can remove the todo and add this to #242 

```suggestion
```

**Comment by larrybabb** [comment]
@korikuzma @jsstevenson @ahwagner could you please try to prioritize reviewing this PR? The primary purpose of it was to get support for the `c.` nomenclature working. I also reverted the hgvs branch dependency that we introduced and will localize our version of it until that hgvs branch is officially merged and a new release is created. I started to consolidate all the hgvs-related logic in the hgvs_tools class, but deferred making any larger strategic design changes until we can meet as a team and agree on what and when we need to do there.

**Comment by korikuzma** [comment]
@larrybabb I am in the middle of reviewing 😄

File: src/ga4gh/vrs/dataproxy.py [hunk 6]
```diff
@@ -88,6 +119,56 @@ def translate_sequence_identifier(self, identifier, namespace=None):
...
         return aliases
+        
+    def derive_refget_accession(self, ac: str):
+        """Derive the refget accession from a public accession identifier
+
+        :param ac: public accession in simple or curie form from which to derive the refget accession
+        :return: Refget Accession if found
+        """
+
+        if ac == None:
```
Comment (korikuzma) [inline src/ga4gh/vrs/dataproxy.py:130]: https://docs.astral.sh/ruff/rules/none-comparison/

```suggestion
        if ac is None:
```

File: src/ga4gh/vrs/dataproxy.py [hunk 7]
```diff
@@ -126,7 +207,7 @@ def _get_sequence(self, identifier, start=None, end=None):
 
     def _get_metadata(self, identifier):
         ns, a = coerce_namespace(identifier).split(":", 2)
-        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))
+        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))
```
Comment (korikuzma) [inline src/ga4gh/vrs/dataproxy.py:210]:
```suggestion
        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))
```

File: src/ga4gh/vrs/extras/translator.py [hunk 8]
```diff
@@ -47,11 +40,12 @@ class Translator:
         re.IGNORECASE
     )
     hgvs_re = re.compile(r"[^:]+:[cgmnpr]\.")
+    hgvs_re = re.compile(r"[^:]+:[cgmnpr]\.")
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:43]: We can remove both since it now lives in `HgvsTools`

```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 9]
```diff
@@ -356,7 +275,8 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
         }
 
         """
-
+        
+        print("gnomad_expr: ", gnomad_expr)
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:279]:
```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 10]
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
         g = m.groupdict()
         assembly_name = kwargs.get("assembly_name", self.default_assembly_name)
         sequence = assembly_name + ":" + g["chr"]
-        refget_accession = self._get_refget_accession(sequence)
+        refget_accession = self.data_proxy.derive_refget_accession(sequence)
         if not refget_accession:
             return None
+        
+        print("refget_accession: ", refget_accession)
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:293]:
```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 11]
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
...
-            raise ValidationError(err_msg)
-
-        seq_ref = models.SequenceReference(refgetAccession=refget_accession)
-        location = models.SequenceLocation(sequenceReference=seq_ref, start=start, end=end)
-        sstate = models.LiteralSequenceExpression(sequence=ins_seq)
-        allele = models.Allele(location=location, state=sstate)
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
+        if kwargs.get("require_validation", True):
+            print("validation check for matching reference sequence bases")
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:303]:
```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 12]
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
...
-        valid_ref_seq, err_msg = self._is_valid_ref_seq(sequence, start, end, ref)
-        if kwargs.get("require_validation", True) and not valid_ref_seq:
-            raise ValidationError(err_msg)
-
-        seq_ref = models.SequenceReference(refgetAccession=refget_accession)
-        location = models.SequenceLocation(sequenceReference=seq_ref, start=start, end=end)
-        sstate = models.LiteralSequenceExpression(sequence=ins_seq)
-        allele = models.Allele(location=location, state=sstate)
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:301]: I have a PR (#375) that simplifies this. I can update it once this PR gets merged.
```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 13]
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
...
+        if kwargs.get("require_validation", True):
+            print("validation check for matching reference sequence bases")
+            # validation check for matching reference sequence bases
+            valid_ref_seq, err_msg = self.data_proxy.is_valid_ref_seq(sequence, start, end, ref)
+            if not valid_ref_seq:
+                raise ValidationError(err_msg)
+            
+        values = {"refget_accession": refget_accession, "start": start, "end": end, "literal_sequence": ins_seq}
+        print("values: ", values)
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:310]:
```suggestion
```

File: src/ga4gh/vrs/extras/translator.py [hunk 14]
```diff
@@ -171,32 +115,12 @@ def translate_to(self, vo, fmt):
...
-
     def _get_hgvs_tools(self):
         """ Only create UTA db connection if needed. There will be one connection per
         translator.
         """
         if self.hgvs_tools is None:
-            uta_conn = hgvs.dataproviders.uta.connect()
-            self.hgvs_tools = HgvsTools(uta_conn)
+            self.hgvs_tools = HgvsTools(self.data_proxy)
         return self.hgvs_tools
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:124]: IMO I think we should use the `lazy_property` decorator 

```python
from ga4gh.vrs.extras.decorators import lazy_property

@lazy_property
def hgvs_tools(self):
    """instantiates and returns an HgvsTools instance"""
    return HgvsTools(self.data_proxy)
```

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 15]
```diff
@@ -1,26 +1,333 @@
...
+    """
+    A convenience class that exposes only the tools needed by vrs-python for working with HGVS (Human Genome Variation Society) notation.
 
-    """ A utility class that provides access to hgvs components that have a UTA database dependency"""
-    def __init__(self, uta_conn):
+    Attributes:
+        parser (hgvs.parser.Parser): The HGVS parser object.
+        uta_conn: The UTA (Universal Transcript Archive) connection object.
+        normalizer: The HGVS normalizer object.
+        variantMapper: The HGVS variant mapper object.
```
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:24]: what about `dp`?

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 16]
```diff
@@ -1,26 +1,333 @@
...
+        # just in time instantiation of the normalizer service, with validation set to True
+        if self.normalizer is None:
+            self.normalizer = hgvs.normalizer.Normalizer(self._fetch_uta_conn(), validate=True)
+        return self.normalizer
+    
+    def _fetch_variantMapper(self):
+        # just in time instantiation of the variantMapper service
+        if self.variantMapper is None:
+            self.variantMapper = hgvs.variantmapper.VariantMapper(self._fetch_uta_conn())
+        return self.variantMapper
```
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:51]: IMO, I think we should just initialize these in the `__init__` method

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 17]
```diff
@@ -1,26 +1,333 @@
...
+    """
+    hgvs_re = re.compile(r"[^:]+:[cgmnpr]\.")
 
-        self.uta_conn = uta_conn
-        self.normalizer = hgvs.normalizer.Normalizer(self.uta_conn)
+    def __init__(self,dp: _DataProxy = None):
         self.parser = hgvs.parser.Parser()
+        self.uta_conn = None
+        self.normalizer = None
+        self.variantMapper = None
```
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:32]:
```suggestion
        self.variant_mapper = None
```

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 18]
```diff
@@ -1,26 +1,333 @@
...
+        Returns:
+            bool: True if the SequenceVariant is intronic, False otherwise.
+        """
+        if isinstance(sv.posedit.pos, hgvs.location.BaseOffsetInterval):
+            return (sv.posedit.pos.start.is_intronic or sv.posedit.pos.end.is_intronic)
+        return False
+    
+
+    def get_edit_type(self, sv: hgvs.sequencevariant.SequenceVariant):
+        if sv == None or sv.posedit == None or sv.posedit.edit == None:
```
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:93]:
```suggestion
        if sv is None or sv.posedit is None or sv.posedit.edit is None:
```

File: src/ga4gh/vrs/utils/hgvs_tools.py [hunk 19]
```diff
@@ -1,26 +1,333 @@
...
+
+        refget_accession = self.dp.derive_refget_accession(sv.ac)
+        if not refget_accession:
+            return None
+        
+        # translate coding coordinates to positional coordinates, if necessary
+        if sv.type == "c":
+            sv = self.c_to_n(sv)
+
+        # TODO: shouldn't we be validating the "ref" sequence here? if the kwargs "require_validation" value is set to True
```
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:181]: This will be added in #151

**Review by korikuzma — CHANGES_REQUESTED** [review]
@larrybabb Great work! Overall, I like this change. I made some suggestions

File: src/ga4gh/vrs/utils/hgvs_tools.py (hunk 19 above)
Comment (korikuzma) [inline src/ga4gh/vrs/utils/hgvs_tools.py:181]:
```suggestion
```

**Comment by jsstevenson** [comment]
@larrybabb as Kori said, I'm out for at least today (which will probably entail some frantic catch-up once I'm back) but for the sake of pragmatism, we can assume my approval once Kori's additional notes above are addressed

File: notebooks/variation_normalizer_rest_dp.py (hunk 5 above)
Comment (larrybabb) [reply notebooks/variation_normalizer_rest_dp.py:4]: yes

File: src/ga4gh/vrs/extras/translator.py (hunk 14 above)
Comment (larrybabb) [reply src/ga4gh/vrs/extras/translator.py:124]: +1

File: src/ga4gh/vrs/extras/translator.py (hunk 12 above)
Comment (larrybabb) [reply src/ga4gh/vrs/extras/translator.py:301]: +1

File: src/ga4gh/vrs/utils/hgvs_tools.py (hunk 16 above)
Comment (larrybabb) [reply src/ga4gh/vrs/utils/hgvs_tools.py:51]: +1

File: src/ga4gh/vrs/utils/hgvs_tools.py (hunk 15 above)
Comment (larrybabb) [reply src/ga4gh/vrs/utils/hgvs_tools.py:24]: done

File: src/ga4gh/vrs/utils/hgvs_tools.py (hunk 17 above)
Comment (larrybabb) [reply src/ga4gh/vrs/utils/hgvs_tools.py:32]: on it

File: src/ga4gh/vrs/extras/translator.py [hunk 20]
```diff
@@ -258,6 +152,28 @@ def __init__(
...
+    def _create_allele(self, values: dict, **kwargs):
+        """
+        Create an allele object with the given parameters.
+
+        Args:
+            values (dict): The values to use for creating the allele object.
+                'refget_accession' (str): The accession ID of the reference genome.
+                'start' (int): The start position of the allele.
+                'end' (int): The end position of the allele.
+                literal_sequence' (str): The literal sequence for the allele.
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:164]:
```suggestion
                'literal_sequence' (str): The literal sequence for the allele.
```

File: src/ga4gh/vrs/extras/translator.py [hunk 21]
```diff
@@ -376,95 +290,22 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
...
-        valid_ref_seq, err_msg = self._is_valid_ref_seq(sequence, start, end, ref)
-        if kwargs.get("require_validation", True) and not valid_ref_seq:
-            raise ValidationError(err_msg)
-
-        seq_ref = models.SequenceReference(refgetAccession=refget_accession)
-        location = models.SequenceLocation(sequenceReference=seq_ref, start=start, end=end)
-        sstate = models.LiteralSequenceExpression(sequence=ins_seq)
-        allele = models.Allele(location=location, state=sstate)
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
```
Comment (korikuzma) [inline src/ga4gh/vrs/extras/translator.py:293]:
```suggestion
```

**Review by korikuzma — APPROVED** [review]
Two minor things. Good job @larrybabb
